logictest: use of span after Finish with high verbosity and plpgsql cursors

pkg/sql/logictest/tests/local/local_test.TestLogic_plpgsql_cursor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/11980159?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/11980159?buildTab=artifacts#/) on master @ [27e7c166b5591b0e05bb2fd23780309b490d7af9](https://github.com/cockroachdb/cockroach/commits/27e7c166b5591b0e05bb2fd23780309b490d7af9):


```
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x6c772b8, 0xc0059fee70})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x3a
github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x6c772b8, 0xc0059fee70}, 0x2, 0x1, 0x0, 0xe0?, {0x58154b3, 0x1e}, {0xc004d0de68, 0x4, ...})
	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:104 +0x4c8
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
github.com/cockroachdb/cockroach/pkg/util/log.Infof(...)
	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:277
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytesLocked(0xc00261c000, {0x6c772b8, 0xc0059fee70}, 0x1?)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:1055 +0x2ac
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytes(0xc004d0df80?, {0x6c772b8?, 0xc0059fee70?}, 0x30?)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:1033 +0xa8
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Close(0x0?, {0x6c772b8?, 0xc0059fee70?})
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:898 +0x45
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).Close(...)
	github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:190
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*DiskBackedRowContainer).Close(0xc0062b4000, {0x6c772b8, 0xc0059fee70})
	github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/row_container.go:591 +0xc5
github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close(0xc0024fe2f0, {0x6c772b8, 0xc0059fee70})
	github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:134 +0x46
github.com/cockroachdb/cockroach/pkg/sql.(*plpgsqlCursorHelper).Close(0xc0024fe2c0)
	github.com/cockroachdb/cockroach/pkg/sql/routine.go:472 +0x78
github.com/cockroachdb/cockroach/pkg/sql.(*cursorMap).closeAll(0xc003f6bd68, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/sql_cursor.go:348 +0xba
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).rollbackSQLTransaction(0xc003f6b980, {0x6c772b8, 0xc00638eff0}, {0x6ca20e0, 0xa1aa938})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1496 +0x65
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc003f6b980, {0x6c772b8, 0xc00638eff0}, {{0x6ca20e0, 0xa1aa938}, {0x0, 0x0, 0x0}, {0xc00248721a, 0x5}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:842 +0x2e9b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x6c772b8?, 0xc005f725d0?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:144 +0xbc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0xc004d0f240?, {0x6c772b8?, 0xc005f725d0?}, {0x6ca20e0?, 0xa1aa938?}, 0x100000000?, 0x1?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3237 +0x32c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc003f6b980, {0x6c772b8, 0xc005f725d0}, {{0x6ca20e0, 0xa1aa938}, {0x0, 0x0, 0x0}, {0xc00248721a, 0x5}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:143 +0x43e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc004d0f988, 0xc004d0f5b0, 0xc003f6b980, 0xc004d0f4e0, 0xc004d0f5d0, 0xc004d0f5c0, 0xc004d0f5e0, 0xc004d0f5f0)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2223 +0x319
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc003f6b980)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2228 +0x3aa
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc003f6b980, {0x6c772b8, 0xc00bce0f60}, 0xc005a05cc8?, 0xedfc06?, 0xc00525bc10?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2145 +0x216
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc00965fc20?, {0x6c772b8?, 0xc00bce0f60?}, {0xc001acd01c?}, 0x4?, 0xc00bce0b40?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:954 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0xc0060ec400, {0x6c772b8, 0xc00bce0f60}, {0x0, 0x4, {0x1, {0x5798641, 0x3}, {0xc003c3e740, 0xf}, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:247 +0x405
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3()
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:999 +0xf5
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:996 +0x668
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_plpgsql_cursor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31923

```
[06:00:55] CREATE OR REPLACE FUNCTION f() RETURNS INT AS $$
  DECLARE
    curs STRING := 'foo';
  BEGIN
    OPEN curs FOR SELECT 1;
    RETURN 0;
  END
$$ LANGUAGE PLpgSQL;
BEGIN;
SELECT f();;
[06:00:55] FETCH FORWARD 3 FROM foo;;
[06:00:55] 	-- OK;
[06:00:55] ABORT;;
*
* ERROR: a SQL panic has occurred while executing the following statement:
* ROLLBACK TRANSACTION
*
*
* ERROR: a panic has occurred!
* use of Span after Finish. Span: udf-stmt-_stmt_open_1-1. Finish previously called at: <stack not captured. Set debugUseAfterFinish>
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:952
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
*   | github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog
*   | 	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92
*   | github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal
*   | 	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:104
*   | github.com/cockroachdb/cockroach/pkg/util/log.logfDepth
*   | 	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
*   | github.com/cockroachdb/cockroach/pkg/util/log.Infof
*   | 	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:277
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytesLocked
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:1055
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytes
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:1033
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Close
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:898
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:190
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*DiskBackedRowContainer).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/row_container.go:591
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:134
*   | github.com/cockroachdb/cockroach/pkg/sql.(*plpgsqlCursorHelper).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/routine.go:472
*   | github.com/cockroachdb/cockroach/pkg/sql.(*cursorMap).closeAll
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sql_cursor.go:348
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).rollbackSQLTransaction
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1496
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:842
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:144
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3237
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:143
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2223
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2228
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2145
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:954
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:247
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:999
* Wraps: (2) while executing: ROLLBACK TRANSACTION
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:952
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | [...repeated from below...]
* Wraps: (4) assertion failure
* Wraps: (5) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
*   | github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog
*   | 	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92
*   | github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal
*   | 	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:104
*   | github.com/cockroachdb/cockroach/pkg/util/log.logfDepth
*   | 	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
*   | github.com/cockroachdb/cockroach/pkg/util/log.Infof
*   | 	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:277
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytesLocked
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:1055
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).releaseBytes
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:1033
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Close
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:898
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:190
*   | github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*DiskBackedRowContainer).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/row_container.go:591
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowContainerHelper).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/buffer_util.go:134
*   | github.com/cockroachdb/cockroach/pkg/sql.(*plpgsqlCursorHelper).Close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/routine.go:472
*   | github.com/cockroachdb/cockroach/pkg/sql.(*cursorMap).closeAll
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sql_cursor.go:348
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).rollbackSQLTransaction
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1496
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:842
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:144
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3237
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:143
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2223
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2228
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2145
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:954
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:247
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:999
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1598
* Wraps: (6) use of Span after Finish. Span: udf-stmt-_stmt_open_1-1. Finish previously called at: <stack not captured. Set debugUseAfterFinish>
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *assert.withAssertionFailure (5) *withstack.withStack (6) *errutil.leafError
*
panic: use of Span after Finish. Span: udf-stmt-_stmt_open_1-1. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: routine_open_cursor-limited: no bytes to release, current 0, free 10240
```

--- a/pkg/sql/routine.go
+++ b/pkg/sql/routine.go
@@ -246,7 +246,7 @@ func (g *routineGenerator) startInternal(ctx context.Context, txn *kv.Txn) (err 
 			w = rrw
 		} else if openCursor {
 			// The result of the first statement will be used to open a SQL cursor.
-			cursorHelper, err = g.newCursorHelper(ctx, plan.(*planComponents))
+			cursorHelper, err = g.newCursorHelper(plan.(*planComponents))
 			if err != nil {
 				return err
 			}
@@ -377,9 +377,7 @@ func (d *droppingResultWriter) Err() error {
 	return d.err
 }
 
-func (g *routineGenerator) newCursorHelper(
-	ctx context.Context, plan *planComponents,
-) (*plpgsqlCursorHelper, error) {
+func (g *routineGenerator) newCursorHelper(plan *planComponents) (*plpgsqlCursorHelper, error) {
 	open := g.expr.CursorDeclaration
 	if open.NameArgIdx < 0 || open.NameArgIdx >= len(g.args) {
 		panic(errors.AssertionFailedf("unexpected name argument index: %d", open.NameArgIdx))
@@ -389,15 +387,17 @@ func (g *routineGenerator) newCursorHelper(
 			"opening an unnamed cursor is not yet supported",
 		)
 	}
+	// Use context.Background(), since the cursor can outlive the context in which
+	// it was created.
 	planCols := plan.main.planColumns()
 	cursorHelper := &plpgsqlCursorHelper{
-		ctx:        ctx,
+		ctx:        context.Background(),
 		cursorName: tree.Name(tree.MustBeDString(g.args[open.NameArgIdx])),
 		resultCols: make(colinfo.ResultColumns, len(planCols)),
 	}
 	copy(cursorHelper.resultCols, planCols)
 	cursorHelper.container.Init(
-		ctx,
+		cursorHelper.ctx,
 		getTypesFromResultColumns(planCols),
 		g.p.ExtendedEvalContextCopy(),
 		"routine_open_cursor", /* opName */

